src/RolloutEvaluator.ts

```diff
@@ -1,571 +1,1116 @@
-  identifier: string;
-
-  /** Email address of the user. */
-  email?: string;
+  if (user.country != null) {
+    result[countryAttribute] = user.country;
+  }
 
-  /** Country of the user. */
-  country?: string;
+  if (user.custom != null) {
+    const wellKnownAttributes: string[] = [identifierAttribute, emailAttribute, countryAttribute];
+    for (const attributeName of Object.keys(user.custom)) {
+      if (wellKnownAttributes.indexOf(attributeName) < 0) {
+        result[attributeName] = user.custom[attributeName];
+      }
+    }
+  }
 
-  /** Custom attributes of the user for advanced targeting rule definitions (e.g. user role, subscription type, etc.) */
-  custom?: { [key: string]: string } = {};
+  return result;
 }
 
-export class RolloutEvaluator implements IRolloutEvaluator {
+export class EvaluateContext {
+  private $userAttributes?: { [key: string]: string } | null;
+  get userAttributes(): { [key: string]: string } | null {
+    const attributes = this.$userAttributes;
+    return attributes !== void 0 ? attributes : (this.$userAttributes = this.user ? getUserAttributes(this.user) : null);
+  }
+
+  private $visitedFlags?: string[];
+  get visitedFlags(): string[] { return this.$visitedFlags ??= []; }
 
-  private readonly logger: LoggerWrapper;
+  isMissingUserObjectLogged?: boolean;
+  isMissingUserObjectAttributeLogged?: boolean;
 
-  constructor(logger: LoggerWrapper) {
+  logBuilder?: EvaluateLogBuilder; // initialized by RolloutEvaluator.evaluate
 
-    this.logger = logger;
+  constructor(
+    readonly key: string,
+    readonly setting: Setting,
+    readonly user: User | undefined,
+    readonly settings: Readonly<{ [name: string]: Setting }>
+  ) {
   }
 
-  evaluate(setting: Setting, key: string, defaultValue: SettingValue, user: User | undefined, remoteConfig: ProjectConfig | null): IEvaluateResult {
-    this.logger.debug("RolloutEvaluator.Evaluate() called.");
+  static forPrerequisiteFlag(key: string, setting: Setting, dependentFlagContext: EvaluateContext): EvaluateContext {
+    const context = new EvaluateContext(key, setting, dependentFlagContext.user, dependentFlagContext.settings);
+    context.$userAttributes = dependentFlagContext.$userAttributes;
+    context.$visitedFlags = dependentFlagContext.visitedFlags; // crucial to use the computed property here to make sure the list is created!
+    context.logBuilder = dependentFlagContext.logBuilder;
+    return context;
+  }
+}
 
-    // A negative setting type indicates a flag override (see also Setting.fromValue)
-    if (setting.type < 0 && !isAllowedValue(setting.value)) {
-      throw new TypeError(
-        setting.value === null ? "Setting value is null." :
-        setting.value === void 0 ? "Setting value is undefined." :
-        `Setting value '${setting.value}' is of an unsupported type (${typeof setting.value}).`);
-    }
+export interface IEvaluateResult {
+  selectedValue: SettingValueContainer;
```

Comment:
Low. 
I don't really like the selectedValue name. In Python and .NET it is called "value".

Reply:
Actually, it's called `selectedValue` in .NET too: https://github.com/configcat/.net-sdk/blob/834bc4df5a88bb3ec41a51096fd66f5a7359f580/src/ConfigCatClient/Evaluation/EvaluateResult.cs#L12

I chose this name because `result.value.value` would have looked weird. But of course, we can change it if you insist.